Polyhedron/demo/Polyhedron/Plugins/IO/3mf_io_plugin.cpp

```diff
@@ -0,0 +1,331 @@
+    // Open file
+    ok = true;
+    std::vector<PointRange> all_points;
+    std::vector<PolygonRange> all_polygons;
+    std::vector<std::string> names;
+    QList<Scene_item*> result;
+    std::vector<std::vector<CGAL::Color> > all_colors;
+    int nb_polylines =
+        CGAL::read_polylines_from_3mf(fileinfo.filePath().toUtf8().toStdString(),
+                                      all_points, all_colors, names);
+    if(nb_polylines < 0 )
+    {
+      ok = false;
+      std::cerr << "Error in reading of meshes."<<std::endl;
+      return result;
+    }
+
+    for(int i=0; i< nb_polylines; ++i)
+    {
+      Scene_polylines_item* pol_item = new Scene_polylines_item();
+      PolylineRange& polylines = pol_item->polylines;
+      polylines.push_back(all_points[i]);
+      pol_item->setName(names[i].data());
+      pol_item->invalidateOpenGLBuffers();
+      CGAL::Color c = all_colors[i].front();
+      pol_item->setColor(QColor(c.red(), c.green(), c.blue()));
+      pol_item->setProperty("already_colord", true);
+      result << pol_item;
+      if(add_to_scene)
+        CGAL::Three::Three::scene()->addItem(pol_item);
+    }
+    all_points.clear();
+    all_colors.clear();
+    names.clear();
+    int nb_point_sets =
+        CGAL::read_point_clouds_from_3mf(fileinfo.filePath().toUtf8().toStdString(),
+                                         all_points, all_colors, names);
+    if(nb_point_sets < 0 )
+    {
+      ok = false;
+      std::cerr << "Error in reading of meshes."<<std::endl;
+      return result;
+    }
+    for(int i=0; i< nb_point_sets; ++i)
+    {
+      Scene_points_with_normal_item* pts_item = new Scene_points_with_normal_item();
+      for(std::size_t j = 0; j < all_points[i].size(); ++j)
+      {
+        pts_item->point_set()->insert(all_points[i][j]);
+      }
+      pts_item->setName(names[i].data());
+      pts_item->invalidateOpenGLBuffers();
+      CGAL::Color c = all_colors[i].front();
+      pts_item->setColor(QColor(c.red(), c.green(), c.blue()));
+      pts_item->setProperty("already_colord", true);
+      result << pts_item;
+      if(add_to_scene)
+        CGAL::Three::Three::scene()->addItem(pts_item);
+    }
+    all_points.clear();
+    names.clear();
+    all_colors.clear();
+    int nb_meshes =
+        CGAL::read_soups_from_3mf(fileinfo.filePath().toUtf8().toStdString(),
+                                  all_points, all_polygons, all_colors, names);
+    if(nb_meshes <0 )
+    {
+      ok = false;
+      std::cerr << "Error in reading of meshes."<<std::endl;
+      return result;
+    }
+    for(int i = 0; i< nb_meshes; ++i)
+    {
+      PolygonRange triangles = all_polygons[i];
+      PointRange points = all_points[i];
+      ColorRange colors = all_colors[i];
+      bool ok = true;
+      if(!PMP::is_polygon_soup_a_polygon_mesh(triangles))
+        ok = PMP::orient_polygon_soup(points, triangles);
+      if(!ok)
+      {
+        std::cerr<<"Object is not orientable. Skipped."<<std::endl;
```

Comment:
That's not true, it simply means that some vertices have been duplicated to avoid non-orientability issues.

Reply:
Which means the input item has been modified, I don't think we want to save it in that case.

Reply:
Save? I see read_soups_.. above. To me it seems to be loading but I might be wrong

Reply:
Oops, that's right